Add withdrawals in my-loans and my-investments

balinskia opened the pull request:
Created withdrawal button and connected it with getline.ts
Created generic withdrawal component which represents single withdrawal.

balinskia committed:
Add withdrawals in my-loans and my-investments

Created withdrawal button and conected it with getline.ts
Created generic withdrawal component.

q3k commented on frontend/src/components/borrow/my-loan-tile/MyLoanTile.vue:
```diff
@@ -20,17 +20,40 @@
+        v-if="withdrawal.isCollateralBackAfterPayback" 
+        :withdrawal="withdrawal"
+        :tokenSymbol="loan.loanTokenSymbol"
+        text="COLLATERAL TO RECEIVE " />
```
`Full collateral after payback:`

q3k commented on frontend/src/components/borrow/my-loan-tile/MyLoanTile.vue:
```diff
@@ -20,17 +20,40 @@
+         v-if="withdrawal.isLoanBackAfterPayback"
+         :withdrawal="withdrawal"
+         :tokenSymbol="loan.loanTokenSymbol"
+         text="THE LOAN WAS PAID BACK. GET PAYBACK" />
```
`Your investment after payback:`

q3k commented on frontend/src/components/borrow/my-loan-tile/MyLoanTile.vue:
```diff
@@ -20,17 +20,40 @@
+      <generic-withdrawal
+         v-if="withdrawal.isCollateralBackAfterDefaulted"
+         :withdrawal="withdrawal"
+         :tokenSymbol="loan.loanTokenSymbol" text="THE LOAN IS DEFAULTED. GET COLLATERAL:" />
```
`Your collateral allowance after defaulting:`

q3k commented on frontend/src/components/borrow/my-loan-tile/MyLoanTile.vue:
```diff
@@ -20,17 +20,40 @@
+        v-if="withdrawal.isCollateralBackAfterCanceled"
+        :withdrawal="withdrawal"
+        :tokenSymbol="loan.loanTokenSymbol"
+        text="THE LOAN HAS NOT RAISED FUNDS ON TIME. GET YOUR COLLATERAL BACK" />
```
`Full collateral after fundraising failed:`

q3k reviewed (changes requested):
Good work so far, but needs some changes. Apart from inline comments - this change also doesn't seem to actually introduce withdrawal buttons in the 'my investments' view..?

balinskia committed:
Fix linter issues

q3k commented on frontend/src/api/utils.ts:
```diff
@@ -21,6 +21,21 @@ export async function getCollateralTokenSymbols(loans: Loan[]): Promise<string[]
   return Promise.all(loans.map((l) => l.parameters.collateralToken.symbol()));
 }
 
+export async function getPossibleWithdrawals(loans: Loan[]): Promise<Withdrawal[][]> {
```
Taking a step back, all of `frontend/src/api` shouldn't have to exist. My approach for now has been to let it live according to what's comfortable to consume by the frontend, hopefully teaching us a bit about usage patterns. Then we can slowly merge such patterns into `getline.ts` so that future users don't have to write such amount of boilerplate again.

See [Desire paths](https://en.wikipedia.org/wiki/Desire_path). :)

balinskia committed:
Add withdrawals to my-investements

After review:
Added getPossibleWithdrawalAmountsByToken.
Withdrawal button text is generated by grouping withdrawals by token symbol.
Changed withdrawals labels.
Moved withdrawals buttons and labels to separate folder.

q3k commented on frontend/src/components/withdrawal/WithdrawalButton.vue:
```diff
@@ -0,0 +1,113 @@
+      }
+
+      let buttonText = format(withdrawalsSummaryPerToken[0]);
+      for (let i = 1; i < withdrawalsSummaryPerToken.length; ++i) {
```
`for (const element of withdrawalsSummaryPerToken) {`

ritave replied:
`withdrawalsSummaryPerToken.forEach()` is also pretty cool

q3k commented on frontend/src/api/invest.ts:
```diff
@@ -52,10 +56,14 @@ export async function getMyInvestments(): Promise<LoanToInvestT[]> {
   let libraryLoans = await api.loansByInvestor(user);
-  return await Promise.all(libraryLoans.map(libraryLoanToViewLoan));
+
+  // We cane have many possible withdrawals per investemnt
```
`can have`

ritave commented on frontend/src/api/invest.ts:
```diff
@@ -56,14 +57,13 @@ export async function getMyInvestments(): Promise<LoanToInvestT[]> {
-  const possibleWithdrawals: Withdrawal[][] = await getPossibleWithdrawals(libraryLoans);
-  const possibleWithdrawalsAmount: BigNumber[][] = await getPossibleWithdrawalsAmounts(possibleWithdrawals);
+  // We cane have many possible withdrawals per investemnt
+  const possibleWithdrawalsAmountsByToken: WithdrawalByTokenT[][] = await getPossibleWithdrawalAmountsByToken(libraryLoans);
```
`possibleWithdrawalsAmountsByToken` -> `possibleWithdrawalAmountsByToken`

ritave commented on frontend/src/api/my-loans.ts:
```diff
@@ -17,9 +16,8 @@ export async function getMyLoans(): Promise<MyLoanT[]> {
-  // We cane have many possile withdrawals per loan
-  const possibleWithdrawals: Withdrawal[][] = await getPossibleWithdrawals(libraryLoans);
-  const possibleWithdrawalsAmount: BigNumber[][] = await getPossibleWithdrawalsAmounts(possibleWithdrawals);
+  // We cane have many possible withdrawals per loan
```
`can have`

ritave commented on frontend/src/api/my-loans.ts:
```diff
@@ -17,9 +16,8 @@ export async function getMyLoans(): Promise<MyLoanT[]> {
-  const possibleWithdrawals: Withdrawal[][] = await getPossibleWithdrawals(libraryLoans);
-  const possibleWithdrawalsAmount: BigNumber[][] = await getPossibleWithdrawalsAmounts(possibleWithdrawals);
+  // We cane have many possible withdrawals per loan
+  const possibleWithdrawalsAmountsByToken: WithdrawalByTokenT[][] = await getPossibleWithdrawalAmountsByToken(libraryLoans);
```
`possibleWithdrawalsAmountsByToken` -> `possibleWithdrawalAmountsByToken`

ritave commented on frontend/src/api/utils.ts:
```diff
@@ -21,21 +22,24 @@ export async function getCollateralTokenSymbols(loans: Loan[]): Promise<string[]
-    withdrawals.map(withdrawalsPerLoan =>
-      Promise.all(withdrawalsPerLoan.map(({ token, amount }) => token.humanize(amount))))
+    withdrawals.map(async withdrawalsPerLoan => {
+      const tokensAmounts = await Promise.all(withdrawalsPerLoan.map(({ token, amount }) => token.humanize(amount)));
```
`tokensAmounts` -> `tokenAmounts`

ritave commented on frontend/src/api/withdrawal.ts:
```diff
@@ -1,16 +1,16 @@
 import { WithdrawalReason, Withdrawal, BigNumber } from './index';
 import API from 'api';
-import { withdrawalT } from 'store/my-loans/types';
+import { withdrawalT, WithdrawalByTokenT } from 'store/my-loans/types';
```
lowercase `withdrawalT`

ritave commented on frontend/src/components/withdrawal/WithdrawalButton.vue:
```diff
@@ -41,19 +43,49 @@ export default {
+        return {
+          tokenSymbol,
+          amount: tokenSymbolToAmountSum[tokenSymbol]
+        }
```
`;`

ritave reviewed (commented):
Minor typos

balinskia committed:
Refactor withdrawal button formatting
